### ISSUE TITLE
Sigmoid multiplication attention in Conv layer

### ISSUE THREAD

**Issue by stas-polukeev** [ultralytics/yolov5#12822]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hey, thanks for the great work!
I found one thing, I don't fully understand. I've converted the model to onnx format and while looking through the graph on Netron, I've noticed this sigmoid multiplication attention in every conv block (on screenshot) 
![image](https://github.com/ultralytics/yolov5/assets/90236898/744311f4-12cd-4dc7-9d6c-c91cf5a8821d)
However, when I inspected the code for conv in models/common.py I do not see anython of that sort

`class Conv(nn.Module):
    # Standard convolution with args(ch_in, ch_out, kernel, stride, padding, groups, dilation, activation)
    default_act = nn.SiLU()  # default activation

    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, d=1, act=True):
        """Initializes a standard convolution layer with optional batch normalization and activation."""
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p, d), groups=g, dilation=d, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity()

    def forward(self, x):
        """Applies a convolution followed by batch normalization and an activation function to the input tensor `x`."""
        return self.act(self.bn(self.conv(x)))

    def forward_fuse(self, x):
        """Applies a fused convolution and activation function to the input tensor `x`."""
        return self.act(self.conv(x))`
        
 At what point does this attention happen or what does it all mean?

**Comment by glenn-jocher** [comment]
@stas-polukeev hey there! 👋

Thanks for the kind words and diving deep into YOLOv5's internals! The "sigmoid multiplication" you're seeing in the graph isn't explicitly coded as a separate component in the `Conv` class in `models/common.py`. Instead, it's a part of the architecture's self-attention mechanism integrated within certain blocks, intended to enhance feature representation by allowing the model to focus on relevant parts of the input. 

This mechanism adjusts the activation output based on learned importance weights, which, when visualized or converted (like to ONNX), might look like a separate "attention" step involving sigmoid functions and multiplications. However, it's effectively interwoven with the convolution and activation operations rather than being a clear-cut, distinct step in the code you've reviewed.

It's great to see your interest in understanding YOLOv5's workings! If you have further questions or need help, feel free to reach out. Happy coding! 🚀

**Comment by stas-polukeev** [comment]
@glenn-jocher thanks for the fast reply!
If it's not explicitly coded in the Conv layers, where can I find it in the code?

**Comment by glenn-jocher** [comment]
@stas-polukeev Glad to help! 😊 The attention mechanism described isn't implemented as a separate, identifiable module titled "attention" within the code, hence the confusion. In YOLOv5, features akin to attention are emergent properties of the network architecture and training process rather than clear-cut, separately coded entities.

If you're looking to understand how these properties emerge, I'd recommend focusing on the interaction between different layers and modules (e.g., CSPDarknet, PANet pathways) and how they process and pass along information. The design of these networks inherently allows for prioritization of certain features over others, similar in effect to attention mechanisms but not explicitly coded as such.

For a deeper dive, examining the forward methods of various classes and how they integrate can be enlightening. Feel free to explore the repository further, and if you have specific questions about certain blocks or pathways, I'm here to help! Happy coding! 🚀

**Comment by stas-polukeev** [comment]
@glenn-jocher  Thanks!

**Comment by glenn-jocher** [comment]
@stas-polukeev You're welcome! If you dive deeper or have any more questions, just let me know. Happy exploring YOLOv5! 😄🚀
